### PR TITLE
DFA-2014: Check if session has expired

### PR DESCRIPTION
**Pull request by HarshVasudev**
Redirecting to /sign-in-again page if session has expired.

### CODE REVIEW

File: express/src/middleware/sessionValidator.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import {NextFunction, Request, Response} from "express";
+
+export default function sessionValidator(req: Request, res: Response, next: NextFunction) {
+    if (req.session.user || req.path === "/" || req.path === "/sign-in" || req.path === "/create/get-email") {
```
Comment (CharlesIC) [inline express/src/middleware/sessionValidator.ts]: We can't keep adding URLs here, this is not a sustainable approach. Instead only add this check to pages that require a session.

File: express/src/controllers/sign-in.ts [hunk 2]
```diff
@@ -133,3 +133,7 @@ export const processEnterMobileForm = async function (req: Request, res: Respons
         active: "sign-in"
     });
 };
+
+export const signInAgain = async function (req: Request, res: Response) {
+    res.render("you-need-to-sign-in-again.njk", {});
```
Comment (CharlesIC) [inline express/src/controllers/sign-in.ts:138]:
```suggestion
    res.render("you-need-to-sign-in-again.njk");
``

We don't need to pass empty options, just omit the param.

File: express/src/views/you-need-to-sign-in-again.njk [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+{% from "govuk/components/input/macro.njk" import govukInput %}
+
+{% set active = "sign-in" %}
+
+{% block pageTitle %}Session Expired.{% endblock %}
+
+{% block mainContent %}
+  <h1 class="govuk-heading-l">Your session has expired, You need to sign in again</h1>
+  <p class="govuk-body"><a class="govuk-link" href="/sign-in">Sign-in</a>.</p>
+{% endblock %}
```
Comment (CharlesIC) [inline express/src/views/you-need-to-sign-in-again.njk:12]: This content doesn't seem correct, can you please check against the content designer's copy.

File: express/src/routes/sign-in.ts [hunk 4]
```diff
@@ -19,6 +20,7 @@ import {processLoginOtpMobile} from "../middleware/sign-in-middleware";
 const router = express.Router();
 
 router.get("/sign-in", showSignInFormEmail);
+router.get("/you-need-to-sign-in-again", signInAgain);
```
Comment (CharlesIC) [inline express/src/routes/sign-in.ts:23]: I believe the URL for the timeout page should be `/session-timeout`. Please check against the design spec.

File: express/src/middleware/sessionValidator.ts (hunk 1 above)
Comment (HarshVasudev) [reply express/src/middleware/sessionValidator.ts]: Yes. I have removed this file in the next commit. So probably this does not need to be addressed.

File: express/src/controllers/sign-in.ts (hunk 2 above)
Comment (HarshVasudev) [reply express/src/controllers/sign-in.ts:138]: Done.

File: express/src/routes/sign-in.ts (hunk 4 above)
Comment (HarshVasudev) [reply express/src/routes/sign-in.ts:23]: Done.

File: express/src/views/you-need-to-sign-in-again.njk (hunk 3 above)
Comment (HarshVasudev) [reply express/src/views/you-need-to-sign-in-again.njk:12]: Done.

File: express/src/controllers/sign-in.ts [hunk 5]
```diff
@@ -133,3 +133,7 @@ export const processEnterMobileForm = async function (req: Request, res: Respons
         active: "sign-in"
     });
 };
+
+export const signInAgain = async function (req: Request, res: Response) {
+    res.render("you-need-to-sign-in-again.njk");
+};
```
Comment (CharlesIC) [inline express/src/controllers/sign-in.ts:139]:
```suggestion
export const sessionTimeout = async function (req: Request, res: Response) {
    res.render("session-timeout.njk");
};
```

Please use the existing `session-timeout` template and delete the duplicate template.

File: express/src/views/you-need-to-sign-in-again.njk [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+{% set active = "sign-in" %}
+
+{% block pageTitle %}Session timeout - GOV.UK Sign In{% endblock %}
+
+{% block mainContent %}
+  <h1 class="govuk-heading-l">You need to sign in again</h1>
+  <p class="govuk-body">To keep your account secure, we’ve signed you out because you did not do anything for 60 minutes.</p>
+  <p class="govuk-body"><a class="govuk-link" href="/sign-in">Sign in again</a>.</p>
+{% endblock %}
```
Comment (CharlesIC) [inline express/src/views/you-need-to-sign-in-again.njk:13]: Duplicate view. This can be removed.

**Review by CharlesIC — COMMENTED** [review]
A small clean-up suggested

File: express/src/controllers/sign-in.ts [hunk 7]
```diff
@@ -113,8 +113,7 @@ export const processEnterMobileForm = async function (req: Request, res: Respons
 
     // eslint-disable-next-line @typescript-eslint/ban-ts-comment
     // @ts-ignore
```
Comment (CharlesIC) [inline express/src/controllers/sign-in.ts:115]: I think we can remove these ignore comments now?

**Review by CharlesIC — COMMENTED** [review]
Looks good now 👍

**Review by CharlesIC — APPROVED** [review]

**Pull request merged**